lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/featurepruning/SparseNetworkOptimizer.java

```diff
@@ -0,0 +1,158 @@
+        if (net.get(0) instanceof SparseAveragedPerceptron) {
+            for (; i < numberclasses; ++i) {
+                SparseAveragedPerceptron sap = (SparseAveragedPerceptron) net.get(i);
+                double wt = sap.getWeightVector().getRawWeights().get(featureindex);
+                double avg = sap.getAveragedWeightVector().getRawWeights().get(featureindex);
+                sum += Math.abs(wt);
+                sum += Math.abs(avg);
+                
+                // if the value is sufficiently large, then we have a good weight and should keep.
+                if (sum > this.threshold)
+                    return true;
+            }
+        } else {
+            for (; i < numberclasses; ++i) {
+                LinearThresholdUnit ltu = (LinearThresholdUnit) net.get(i);
+                double wt = ltu.getWeightVector().getRawWeights().get(featureindex);
+                sum += Math.abs(wt);
+                
+                // if the value is sufficiently large, then we have a good weight and should keep.
+                if (sum > this.threshold)
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * In this case, we must check, for each feature, the associated set of weight in each weight
+     * vector, if they are all very small, it is useless. The array returned is sorted ascending.
+     * @return the set of useless features.
+     * @see edu.illinois.cs.cogcomp.lbjava.learn.featurepruning.LexiconOptimizer#identifyUselessFeatures()
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    protected int[] identifyUselessFeatures() {
+        Lexicon lex = networkLearner.demandLexicon();
+        if (lex != null) {
+            HashSet<Feature> whitelist = compileWhitelist(lex);
+
+            // we have the conjunctive features, if left, right, or the parent itself has a non zero weight,
+            // consider non of the features (parent, left or right) useless, whitelist them.
+            int count = 0;
+            int numberfeatures = lex.size();
+            int numberclasses = networkLearner.getNetwork().size();
+            int[] all = new int[numberfeatures];
+            TIntHashSet defunct = new TIntHashSet();
+            for (Object e : lex.getMap().entrySet()) {
+                Entry<Feature, Integer> entry = (Entry<Feature, Integer>) e;
+                int fi = entry.getValue();
+                if (!whitelist.contains(entry.getKey())) {
+                    int i = 0;
+                    for (; i < numberclasses; ++i) {
+                        LinearThresholdUnit ltu = (LinearThresholdUnit) networkLearner.getNetwork().get(i);
+                        if (ltu == null) {
+                            System.out.println("THERE WAS NO LTU AT " + i);
+                            continue;
+                        }
+                        double wt = ltu.getWeightVector().getRawWeights().get(fi);
+
```

Comment:
Ya, this is wrong. We need to call hasWeight here I think, this won't work for SparseAveragedPerceptron where we need to sum the past average with the actual weight.

Reply:
Also, it doesn't even address absolute zero.